[imgui] sdl3 bindings

constcuriosity opened the pull request:
Adds new sdl3-binding and sdl3-renderer-binding features to the imgui port.

Imgui already has backends implemented for sdl3. They just hadn't yet been exposed as options like the sdl2 backends were. This change mainly just copies the same operations that are done for sdl2 support, but swaps the 2 for a 3.

- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [x] SHA512s are updated for each updated download.
- [x] The "supports" clause reflects platforms that may be fixed by this new version.
- [x] Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.
- [x] Any patches that are no longer applied are deleted from the port's directory.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is added to each modified port's versions file.

constcuriosity committed:
Added sdl3 features to the imgui port

constcuriosity committed:
Update imgui versions for sdl3 features

constcuriosity commented:
@microsoft-github-policy-service agree

JavierMatosD commented:
Hi @constcuriosity, thank you for the contribution. Unfortunately, all feature sets in ports should be side by side installable. This is implementing alternatives as features. See -> https://learn.microsoft.com/en-us/vcpkg/contributing/maintainer-guide#do-not-use-features-to-implement-alternatives

Currently, `vcpkg install imgui[sdl2-binding,sdl3-binding]` will not build. 

Marking this vcpkg-team-review for a path forward.

constcuriosity commented:
Hi @JavierMatosD, thanks for taking a look at this pull request. I'll take a look at seeing if there's a way forward for these two features to compile together.

Do you know offhand if all of the directx version features are able to compile together? I'm just looking to extend the pattern that's already been established by the port.

JavierMatosD commented:
>Hi @JavierMatosD, thanks for taking a look at this pull request. I'll take a look at seeing if there's a way forward for these two features to compile together.

I spoke with my coworkers and it seems that the SDL project currently calls 2.x "current". See -> https://www.libsdl.org/. We think SDL2 should remain the canonical one for now. We encourage you to create an overlay port with this added functionality for your own purposes. See -> https://learn.microsoft.com/vcpkg/concepts/overlay-ports

> Do you know offhand if all of the directx version features are able to compile together? I'm just looking to extend the pattern that's already been established by the port.

 These features were probably added erroneously. 
 
I'm closing this PR for now. Please feel free to reopen the PR if you feel we are mistaken.

constcuriosity commented:
> I spoke with my coworkers and it seems that the SDL project currently calls 2.x "current". See -> https://www.libsdl.org/. We think SDL2 should remain the canonical one for now. We encourage you to create an overlay port with this added functionality for your own purposes. See -> https://learn.microsoft.com/vcpkg/concepts/overlay-ports

SDL3 has hit API and ABI lock as listed at https://wiki.libsdl.org/SDL3/FrontPage as well as in the formation of the sdl3 port a bit ago https://github.com/microsoft/vcpkg/pull/40867. While not official yet, projects are now encouraged to move towards adoption of SDL3.

constcuriosity committed:
Merge branch 'master' into imgui-sdl3-bindings

# Conflicts:
#	ports/imgui/vcpkg.json
#	versions/baseline.json
#	versions/i-/imgui.json

constcuriosity committed:
Fixup imgui vcpkg.json post merge

constcuriosity committed:
Updating versions for imgui port

constcuriosity commented:
While you can have a vcpkg.json file that includes both sdl2 and sdl3, you can't follow the usage lines for both of them in your own cmake files. Doing so will have the 2nd library inclusion complain about a difference between the `INTERFACE_SDL_VERSION` and the `SDL_VERSION` of the next to-be-included library. And this is what causes the sdl2-binding and sdl3-binding incompatibility. Those features request either sdl2 or sdl3 as dependencies, and turning on both features will attempt to enforce that both are included, which triggers the above error.

This feels like it breaks the spirit of [Ports in the current baseline must be installable simultaneously](https://learn.microsoft.com/en-us/vcpkg/contributing/maintainer-guide#ports-in-the-current-baseline-must-be-installable-simultaneously), since while you can technically install both SDL2 and SDL3 simultaneously, you can't have both of them as a dependency. If that was possible, then these features would work fine together.

Practically though... I don't know when that'd really happen. I think the vast majority of projects would only ever use one or the other. And same would go for these features on the imgui port. I guess the workaround would be to remove the dependency on each of the features, and rely on the project author add the dependencies themselves, but that's counter to the goal of vcpkg.

So with the TL;DR "It's not the features' fault, it's SDL2 and SDL3 incompatibility", I'd like to request an exception to the "no mutually exclusive features" policy. Making an overlay port would certainly work but it introduces a continuous maintenance burden by forking a frequently updated port (the version was even bumped within the short lifetime of this PR), and I believe the likelihood of a project purposefully including both SDL versions and both features to be very small.

Thanks for your time.

constcuriosity commented:
@JavierMatosD Sorry to ping you directly but it doesn't look like I have permission to re-open the PR myself.

constcuriosity committed:
Merge branch 'master' into imgui-sdl3-bindings

constcuriosity committed:
[imgui] Favor sdl3 dependencies of sdl2

SDL2 and SDL3 are mutually exclusive ports. They're two versions of the same library with API differences. While the ports can be successfully installed alongside each other, you can't include both as dependencies. The SDL library has a version enforcement mechanism.

The best practice though is for vcpkg ports and features to always be installable alongside each other. But we can't do that because of above dependency conflicts. So this commit makes it so that if both the sdl2 and sdl3 features are enabled, the sdl2 ones are effecitvely ignored and a warning is logged to the user.

constcuriosity committed:
[imgui] updating version

constcuriosity commented:
I've authored a compromise where the sdl2-binding, sdl3-binding, sdl2-renderer-binding, and sdl3-renderer-binding features can all be included as enabled features. When both SDL2 and SDL3 features are enabled, the SDL2 features are basically ignored and a warning is logged that says that the SDL2 dependency was not enforced.

So the ports and features now technically satisfy the maintainer guidelines, but I wouldn't say it adheres to the spirit of them. Nevertheless it continues with my statement about that including both SDL2 and SDL3 features is not expected and very much not desired behavior.

dg0yt commented:
The problem is not with projects, but with ports in vcpkg. AFAIU the conflicting features will cause failures in vcpkg CI as soon as one port depends on feature sdl2 and another oner depends on feature sdl3.

BillyONeal commented:
Please note I am speaking personally, not as 'the maintainers' here.

I don't believe we have the power to grant 'exceptions' here, as the rule comes from technical realities of the situation. We can't build the `2^N` universes of dependencies allowing effectively mutually exclusive features would force. We have to choose a reasonable answer to what is the canonical vcpkg answer in such cases, and users who want something else have to leave the bounds of the curated registry.

> The problem is not with projects, but with ports in vcpkg. AFAIU the conflicting features will cause failures in vcpkg CI as soon as one port depends on feature sdl2 and another oner depends on feature sdl3.

In addition to what dg0yt says here, I want to note that this is in large part what vcpkg is trying to 'sell'. That you can show up to our curated registry, do `vcpkg install a b` for any `a` and `b` in that registry, and the resulting thing be functional. Our CI is how we *test* that that works, but it is more than just an infrastructure concern for us.

As long as SDL's owners consider and document SDL2 as the current version, I believe as far as our curated registry is concerned, that needs to be the version ports use when there are questions like this.

SDL2 and SDL3 are in the small group of major version breaks where the upstream developers go out of their way to make the break 'clean' and simultaneously installable. The big example in our registry that also did this was qt5 and qt6. For projects that did not implement the separation qt itself did, like `qt-advanced-docking-system`, we held the canonical version that downstream port used at qt5 for some time after qt6's release. It seems unlikely to me that the `imgui` is any different than `qt-advanced-docking-system` here.

>SDL3 is just about to become the stable version. If you look at SDL's [GitHub](https://github.com/libsdl-org/SDL) repository, you will see that they already switched the main branch to SDL3 and moved SDL2 into a separate branch.

It makes sense that their development repo would be a leading indicator on what they consider to be the canonical version. vcpkg, because we have to make sure downstream dependencies work when that switch is flipped, will be lagging.

Agorath commented:
Thanks for the detailed explanation! After considering it, I agree that you are correct, and I'd like to withdraw my "application" for an exception. 😅
The only other viable course of action, if we wanted to have both features be compatible, would have to be to ask the SDL developers or submit a pull request to adjust their CMake scripts so that SDL2 and SDL3 can be built simultaneously. I looked at what causes the failure, and it all seems to come down to the variable `SDL_VERSION` being used in both versions' CMake scripts, which leads to the build failure. If this variable were renamed to `SDL_VERSION{2/3}` respectively, I surmise that the build would work.
I'll try that with local forks of both SDL2 and 3 or with a vcpkg patch in a custom registry to see if that solves the issue. If it does, I'll submit a pull request to the SDL repo or vcpkg, depending on which solution works better.

Agorath commented:
I'd like to say I'm not exactly well-versed in CMake and only scrape by with a healthy dose of intuition when working with it.

Okay, that being said, I looked at the respective CMakeLists.txt files, and to me, modifying SDL_VERSION to be SDL{2/3]_VERSION wouldn't be near enough to fix this problem. There are many other variables in the respective CMake files that share the same name, i.e. SDL_OPENGL and many, many more, which could run into the same problems.

That said, being able to use the ImGui vcpkg port for either SDL2 or SDL3 will be a requirement for the foreseeable future. Many legacy apps use SDL2, and not exposing the bindings for SDL3 in general, and its new GPU API in particular, will be a problem for new projects or projects that want to upgrade.

Would anyone happen to have any ideas or input on the above? 🤔 (Particularly the bits about CMake)

In the meantime, I'll open an issue with the SDL project and ask for their opinion on this problem.

Cheney-W commented:
@vicroms Please take a look.

JavierMatosD commented:
> We wait until SDL3 is officially released and then drop SDL2 for SDL3 in the imgui port as proposed above.
=> This would, from that point onwards, prevent users with legacy SDL2 applications from upgrading to newer imgui versions via vcpkg.

After discussing this with @vicroms, @ras0219-msft, @AugP, @BillyONeal we believe that this is the right approach. Please more this out of draft when SDL3 is officially released.

scschaefer commented:
Okay, thanks for the feedback. 🙂 

I don't know if the following is possible under the vcpkg rules, but wouldn't it be possible to offer an "imgui-sdl2" port from that point in time when we switch the main imgui port to sdl3?
So, we would, more or less, make a copy of the current port with a new name before the SDL3 switch. This way, existing SDL2 users would only have to switch to that new "backwards compatibility" port to continue to benefit from imgui bug fixes and improvements. 🤔

Agorath commented:
Okay, thanks for the feedback. 🙂

I don't know if the following is possible under the vcpkg rules, but wouldn't it be possible to offer an "imgui-sdl2" port from that point in time when we switch the main imgui port to sdl3?
So, we would, more or less, make a copy of the current port with a new name before the SDL3 switch. This way, existing SDL2 users would only have to switch to that new "backwards compatibility" port to continue to benefit from imgui bug fixes and improvements. 🤔

MHebes commented:
Incredibly, SDL3 (3.2.0) has been officially released! https://www.patreon.com/posts/120491416

MasterDrake commented:
So, what's the next move or general consensus of what to do?

constcuriosity committed:
Merge branch 'master' into imgui-sdl3-bindings

# Conflicts:
#	ports/imgui/vcpkg.json
#	versions/baseline.json
#	versions/i-/imgui.json
- Incremented port version imgui/vcpkg.json
- Updated baseline.json with updated port version
- Too master's version of imgui.json. Going to need to regenerate version.

constcuriosity committed:
Updating imgui.json port version

dg0yt commented:
> > We wait until SDL3 is officially released and then drop SDL2 for SDL3 in the imgui port as proposed above.
> > => This would, from that point onwards, prevent users with legacy SDL2 applications from upgrading to newer imgui versions via vcpkg.
> 
> After discussing this with @vicroms, @ras0219-msft, @AugP, @BillyONeal we believe that this is the right approach. Please more this out of draft when SDL3 is officially released.

Reminder, sdl2 features are expected to be removed at this point.

Agorath commented:
Hi, sorry as well for my radio silence the last few weeks... Work work...
I was about to say that the changes of commit [ee88b23](https://github.com/microsoft/vcpkg/pull/42850/commits/ee88b239a665458a825a46966eeef20358bf0ea1) would have to be reversed and all references to SDL2 would have to be removed before continuing. 🤔

constcuriosity commented:
>> We wait until SDL3 is officially released and then drop SDL2 for SDL3 in the imgui port as proposed above.
=> This would, from that point onwards, prevent users with legacy SDL2 applications from upgrading to newer imgui versions via vcpkg.

> After discussing this with @vicroms, @ras0219-msft, @AugP, @BillyONeal we believe that this is the right approach. Please more this out of draft when SDL3 is officially released.

Alright... That doesn't bother my side project because I'm starting fresh with sdl3 and haven't used sdl2 before, but what's the recourse for anyone that was using sdl2 before? Are they expected to maintain their own overlay port at that point? Just want it to be very explicit for anyone that needs to deal with the downstream wake.

Also, it looks like because I did a merge on some of the files instead of a complete stomp and edit, it's blocking the pull request completely. Is there a way to salvage the pull, steam roll everything, or is it just cleaner to make a fresh pull request? Not really a fan of losing the context of this thread though.

dg0yt commented:
> Also, it looks like because I did a merge on some of the files instead of a complete stomp and edit, it's blocking the pull request completely. Is there a way to salvage the pull, steam roll everything, or is it just cleaner to make a fresh pull request? Not really a fan of losing the context of this thread though.

The "Merging is blocked ... This branch must not contain merge commits" also appears on clean PRs ATM. I suggest to ignore it.

constcuriosity committed:
[imgui] Removed SDL2 feature bindings

Since SDL2 and SDL3 can't coexist, and vcpkg is designed to support the combination of any and all ports and features, the SDL2 and SDL3 features can't exist at the same time. Now that SDL3 is official, the decision of vcpkg team is to remove the old SDL2 bindings to support these new SDL3 ones.

constcuriosity committed:
[imgui] Updated imgui port version

Agorath commented:
> Alright... That doesn't bother my side project because I'm starting fresh with sdl3 and haven't used sdl2 before, but what's the recourse for anyone that was using sdl2 before? Are they expected to maintain their own overlay port at that point? Just want it to be very explicit for anyone that needs to deal with the downstream wake.
I'll work on a new vcpkg port for the new "sdl2-compat"-layer I mentioned above. I plan to create a pull request after the sdl2-compat port is approved and merged, which adds the sdl2 feature back to the imgui port via the sdl2-compat layer. This way, users with existing sdl2 projects should still be able to continue using recent versions of imgui.

constcuriosity commented:
Ok, latest changes remove the SDL2 feature bindings and replaces them with the SDL3 ones.

Agorath reviewed (approved):
The changes look good to me.
The previous entries for SDL2 were replaced with SDL3.

Cheney-W commented:
New feature test passed with below triplets:
```
x86-windows
x64-windows
x64-windows-static
```

MasterDrake commented:
Finally, big thanks to everyone!
can't wait to actually use sdl3 :D
